### DISCUSSION TITLE
How to link again the unlink users in Keycloak User Federation?

### DISCUSSION THREAD

**Discussion by kresnasatya** [keycloak/keycloak#10490] [Q&A] [answered]
Hi, I link my LDAP users to Keycloak in `WRITEABLE` edit mode. I see field `federation link` in my user edit form. I accidentally press the "Unlink Users" button, I check there's no `federation link` in my user edit form. When, I want to login the user with username and password it doesn't successfully login. It says that "username or passsword are not correct".

I check in Keycloak docs that when user login in User Federation LDAP way then Keycloak will validate password directly to LDAP server. So, when I unlink users then the connection to validate password disabled.

My question is there Keycloak provide a way to link again the unlink users in Keycloak User Federation? Also, why there's no confirmation like modal or alert that says "If you unlink users then you cannot link again the unlink users and your user cannot login because Keycloak cannot validate password to your LDAP server. Are you sure?".

I also see the same issue that mentioned in StackOverflow: https://stackoverflow.com/questions/54615957/keycloak-gives-username-or-password-is-wrong-error-after-pressed-unlink-users.

<img width="1436" alt="Screen Shot 2022-02-26 at 14 37 08" src="https://user-images.githubusercontent.com/11099186/155868423-cbb52b28-ae8d-42ae-adb2-89027bb19a36.png">

<img width="1552" alt="Screen Shot 2022-02-27 at 12 13 53" src="https://user-images.githubusercontent.com/11099186/155868317-403a61d9-9a1c-4089-8b85-3283675c98b4.png">

<img width="1436" alt="Screen Shot 2022-02-26 at 14 36 22" src="https://user-images.githubusercontent.com/11099186/155868416-6639c332-4d28-467e-b6cc-ad7e0244c802.png">

**Comment by bs-matil** [comment]
If its an option (u have no other state on it) the deletion of the user will fix your issue

**Reply by kresnasatya** [reply]
Yeah, I think that’s the last option that I have. But, imagine if I use Keycloak more than years, have more than 100k users from LDAP that already sync it with Keycloak in the same config above, the users also mapped with Keycloak Client and Roles with users id from Keycloak not LDAP. 

Then, suddenly accidentally press “unlink users” which make users cannot login because Keycloak validate password from LDAP server. It makes me to remove the users and sync again and mapped client and roles from scratch because users id from Keycloak is different from the previous.

It would be better if Keycloak provides:

1. An alert when someone press “unlink users” button that give a message that it will be unlink the users so if your users login then Keycloak cannot do that because Keycloak cannot connect to LDAP to validate the password from LDAP.
2. An option to “link again the unlink users” to solved the validate password from LDAP.

**Comment by edmaputra** [comment, marked as answer]
Hi @satyakresna 
I have a similar case with you and found the way to restore the unlink user to be the linked one again. I do it manually via the database

1. Go to **COMPONENT** table, and find the configuration name of your LDAP configuration in User Federation. Copy the ID
2. Go to your **USER_ENTITY** table and query the account you want to restore
3. On the **FEDERATION_LINK** column of the account data, **paste the COMPONENT id**
4. Verify by hitting the **Synchronized all user** button in your User Federation configuration page

Hope this help

**Reply by kresnasatya** [reply]
Thanks @edmaputra. I also use this method too. Here's what I do.

Find Federation ID from COMPONENT table based on REALM_ID with query: `SELECT id FROM COMPONENT WHERE NAME = '<mapper-ldap-name>' AND REALM_ID = '<realm-name>';`.
After that, update federation_link column in `USER_ENTITY` based on REALM_ID with query:
`UPDATE USER_ENTITY SET FEDERATION_LINK = '<federation-id>' WHERE REALM_ID = '<realm-name>';`
